pubsub/pubsub.go

```diff
@@ -0,0 +1,272 @@
+		}
+	}
+	b := bundler.NewBundler(msgErrChan{}, handler)
+	if opts.SendDelay == 0 {
+		b.DelayThreshold = SendDelayDefault
+	} else {
+		b.DelayThreshold = opts.SendDelay
+	}
+	if opts.BatchSize != 0 {
+		b.BundleCountThreshold = opts.BatchSize
+	}
+	t := &Topic{
+		driver:  d,
+		batcher: b,
+	}
+	return t
+}
+
+// Subscription receives published messages.
+type Subscription struct {
+	driver driver.Subscription
+
+	// ackBatcher makes batches of acks and sends them to the server.
+	ackBatcher *bundler.Bundler
+
+	// q is the local queue of messages downloaded from the server.
+	sem chan struct{}
+	q   []*Message
+}
+
+// SubscriptionOptions contains configuration for Subscriptions.
+type SubscriptionOptions struct {
+	// AckDelay tells the max duration to wait before sending the next batch
+	// of acknowledgements back to the server.
+	AckDelay time.Duration
+
+	// AckBatchSize is the maximum number of acks that should be sent to
+	// the server in a batch.
+	AckBatchSize int
+
+	// AckDeadline tells how long the server should wait before assuming a
+	// received message has failed to be processed.
+	AckDeadline time.Duration
+}
+
+// AckDelayDefault is the value for SubscriptionOptions.AckDelay if it is not set.
+const AckDelayDefault = time.Millisecond
+
+// Receive receives and returns the next message from the Subscription's queue,
+// blocking and polling if none are available. This method can be called
+// concurrently from multiple goroutines. The Ack() method of the returned
+// Message has to be called once the message has been processed, to prevent it
+// from being received again.
+func (s *Subscription) Receive(ctx context.Context) (*Message, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-s.sem:
+	}
+	defer func() {
+		s.sem <- struct{}{}
+	}()
+	if len(s.q) == 0 {
+		if err := s.getNextBatch(ctx); err != nil {
+			return nil, err
+		}
+	}
+	m := s.q[0]
+	s.q = s.q[1:]
+	return m, nil
+}
+
+// getNextBatch gets the next batch of messages from the server.
+func (s *Subscription) getNextBatch(ctx context.Context) error {
+	msgs, err := s.driver.ReceiveBatch(ctx)
+	if err != nil {
+		return err
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		if len(msgs) == 0 {
+			return errors.New("subscription driver bug: received empty batch")
+		}
+		s.q = make([]*Message, len(msgs))
+		for i, m := range msgs {
+			s.q[i] = &Message{
+				Body:       m.Body,
+				Attributes: m.Attributes,
+				ackID:      m.AckID,
+				sub:        s,
+			}
+		}
+	}
+	return nil
+}
+
+// Close flushes pending ack sends and disconnects the Subscription.
+func (s *Subscription) Close() error {
+	s.ackBatcher.Flush()
+	return s.driver.Close()
+}
+
+// NewSubscription creates a Subscription from a driver.Subscription and opts to
+// tune sending and receiving of acks and messages. Behind the scenes,
+// NewSubscription spins up a goroutine to gather acks into batches and
+// periodically send them to the server.
```

Comment:
Similar to `blob`, add a comment like `// It is for use by provider implementations.`.

(this is public in the package so that driver impls can call it, but will not be called by end users).

Reply:
Done.